File: py110/small_problems_110_119/advanced/6.py
# https://launchschool.com/exercises/97414944
"""

input: sorted list, item to be found
output: index of found element or -1 if not found
rules:


Algo:
1. left = 0 and a right = len(list) - 1
2. while left < right:
3. find the mid index. and then check if mid index element is your answer if so return mid
4. if not then you compare if the number is greater then the mid. 
5. if it is then you make right = to mid + 1
6. else if the number is less then the mid then left = mid - 1
7 return -1

"""

'''

3
l = 2
r = 2

m = (2 + 2) // 2 = 2

012
123

'''

def binary_search(lst, item):
    left, right = 0, len(lst) - 1

    while left <= right:
        mid = (left + right) // 2
        if lst[mid] == item:
            return mid
        if item > lst[mid]:
            left = mid + 1
        else:
            right = mid - 1            
    return -1

# All of these examples should print True
businesses = ['Apple Store', 'Bags Galore', 'Bike Store',
              'Donuts R Us', 'Eat a Lot', 'Good Food',
              'Pasta Place', 'Pizzeria', 'Tiki Lounge',
              'Zooper']
print(binary_search(businesses, 'Pizzeria') == 7)
print(binary_search(businesses, 'Apple Store') == 0)

print(binary_search([1, 5, 7, 11, 23, 65, 89, 102], 77) == -1)
print(binary_search([1, 5, 7, 11, 23, 65, 89, 102], 89) == 6)
print(binary_search([1, 5, 7, 11, 23, 65, 89, 102], 5) == 1)

names = ['Alice', 'Bonnie', 'Kim', 'Pete', 'Rachel', 'Sue',
         'Tyler']
print(binary_search(names, 'Peter') == -1)
print(binary_search(names, 'Tyler') == 6)

File: ls215/smallProblems/adv/6.js
// https://launchschool.com/exercises/26cebb22

/*

c = 89

[1, 5, 7, 11, 23, 45, 65, 89, 102
 0.  1 2.  3.  4.  5.  6. 7.   8

l = 0
h = 9

mid = Math.floor((l + h) / 2)

mid = 0 + 9 / 2 = 4





*/


'use strict';

function binarySearch(list, candidate) {
  let low = 0;
  let high = list.length;

  while (low <= high) {
    let mid = Math.floor((low + high) / 2);

    if (list[mid] === candidate) {
      return mid;
    } else if (list[mid] < candidate) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  return -1;
}



const yellowPages = ['Apple Store', 'Bags Galore', 'Bike Store', 'Donuts R Us', 'Eat a Lot', 'Good Food', 'Pasta Place', 'Pizzeria', 'Tiki Lounge', 'Zooper'];
console.log(binarySearch(yellowPages, 'Pizzeria'));                   // 7
console.log(binarySearch(yellowPages, 'Apple Store'));                // 0

console.log(binarySearch([1, 5, 7, 11, 23, 45, 65, 89, 102], 77));    // -1
console.log(binarySearch([1, 5, 7, 11, 23, 45, 65, 89, 102], 89));    // 7
console.log(binarySearch([1, 5, 7, 11, 23, 45, 65, 89, 102], 5));     // 1

console.log(binarySearch(['Alice', 'Bonnie', 'Kim', 'Pete', 'Rachel', 'Sue', 'Tyler'], 'Peter'));  // -1
console.log(binarySearch(['Alice', 'Bonnie', 'Kim', 'Pete', 'Rachel', 'Sue', 'Tyler'], 'Tyler'));  // 6